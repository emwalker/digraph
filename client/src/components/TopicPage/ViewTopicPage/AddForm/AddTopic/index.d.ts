import React from 'react';
import { AddTopic_viewer$key } from '__generated__/AddTopic_viewer.graphql';
import { AddTopic_parentTopic$key } from '__generated__/AddTopic_parentTopic.graphql';
type Props = {
    disabled?: boolean;
    parentTopic: AddTopic_parentTopic$key;
    viewer: AddTopic_viewer$key;
};
export default function AddTopic(props: Props): React.JSX.Element | null;
export {};
