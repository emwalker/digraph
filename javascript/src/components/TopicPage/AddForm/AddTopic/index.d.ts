import { RelayProp } from 'react-relay';
import { AddTopic_viewer$data as Viewer } from '__generated__/AddTopic_viewer.graphql';
import { AddTopic_topic$data as Topic } from '__generated__/AddTopic_topic.graphql';
declare type Props = {
    disabled?: boolean;
    relay: RelayProp;
    topic: Topic;
    viewer: Viewer;
};
declare const _default: import("react-relay").Container<Pick<Omit<Props, "relay">, "topic" | "viewer"> & {
    disabled?: boolean | undefined;
} & {}>;
export default _default;
