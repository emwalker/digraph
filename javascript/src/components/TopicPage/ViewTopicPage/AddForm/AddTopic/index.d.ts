import { AddTopic_viewer$key } from '__generated__/AddTopic_viewer.graphql';
import { AddTopic_topic$key } from '__generated__/AddTopic_topic.graphql';
declare type Props = {
    disabled?: boolean;
    topic: AddTopic_topic$key;
    viewer: AddTopic_viewer$key;
};
export default function AddTopic(props: Props): JSX.Element;
export {};
