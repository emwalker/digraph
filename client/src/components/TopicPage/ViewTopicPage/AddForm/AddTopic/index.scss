.form-container {
  text-align: right;
}
