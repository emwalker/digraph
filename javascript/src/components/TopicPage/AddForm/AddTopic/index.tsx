import React, { Component, KeyboardEvent, FormEvent } from 'react'
import { graphql, createFragmentContainer, RelayProp, DeclarativeMutationConfig } from 'react-relay'

import upsertTopicMutation from 'mutations/upsertTopicMutation'
import { AddTopic_viewer as Viewer } from '__generated__/AddTopic_viewer.graphql'
import { AddTopic_topic as Topic } from '__generated__/AddTopic_topic.graphql'

type RepositoryType = Viewer['selectedRepository']

const tooltipText = 'Add a subtopic to this topic. You can click "Edit"\n'
  + 'afterwards if it also belongs under another topic.\n'
  + 'Press "Return" to submit the new topic.'

type Props = {
  disabled?: boolean,
  relay: RelayProp,
  topic: Topic,
  viewer: Viewer,
}

type State = {
  name: string,
}

class AddTopic extends Component<Props, State> {
  static defaultProps = {
    disabled: true,
  }

  constructor(props: Props) {
    super(props)
    this.state = {
      name: '',
    }
  }

  onKeyPress = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') this.createTopic()
  }

  get selectedRepo(): RepositoryType {
    return this.props.viewer.selectedRepository
  }

  get orgLogin(): string | null {
    const repo = this.selectedRepo
    if (!repo) return null
    return repo.organization.login
  }

  get relayConfigs(): DeclarativeMutationConfig[] {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_childTopics',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'topicEdge',
    }]
  }

  updateName = (event: FormEvent<HTMLInputElement>) => {
    this.setState({ name: event.currentTarget.value })
  }

  createTopic() {
    const repo = this.selectedRepo
    if (!repo || !this.orgLogin) {
      // eslint-disable-next-line no-console
      console.log('Missing org name or repo, unable to create topic')
      return
    }

    upsertTopicMutation(
      this.props.relay.environment,
      {
        name: this.state.name,
        repositoryName: repo.name,
        organizationLogin: this.orgLogin,
        topicIds: [this.props.topic.id],
      },
      {
        configs: this.relayConfigs,
      },
    )
    this.setState({ name: '' })
  }

  render = () => (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-ne"
          aria-label={tooltipText}
        >
          <label htmlFor="create-topic-name">Add subtopic</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control test-topic-name input-sm"
          disabled={this.props.disabled}
          id="create-topic-name"
          onChange={this.updateName}
          onKeyPress={this.onKeyPress}
          placeholder="Name or description"
          value={this.state.name}
        />
      </dd>
    </dl>
  )
}

export default createFragmentContainer(AddTopic, {
  viewer: graphql`
    fragment AddTopic_viewer on User {
      selectedRepository {
        name
        organization {
          login
        }
      }
    }
  `,
  topic: graphql`
    fragment AddTopic_topic on Topic {
      id
    }
  `,
})
