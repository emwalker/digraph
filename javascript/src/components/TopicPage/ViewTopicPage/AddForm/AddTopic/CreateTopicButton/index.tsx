import React from 'react'

export default function CreateTopicButton() {
  return (
    <button className="btn btn-outline" type="button">
      Create new topic
    </button>
  )
}
