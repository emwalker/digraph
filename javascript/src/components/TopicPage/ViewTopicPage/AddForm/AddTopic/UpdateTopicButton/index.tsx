import React from 'react'

export default function MergeTopicsButton() {
  return (
    <button className="btn btn-primary" type="button">
      Update existing topic
    </button>
  )
}
