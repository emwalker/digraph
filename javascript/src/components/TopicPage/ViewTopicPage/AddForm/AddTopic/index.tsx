import React, { FormEvent, KeyboardEvent, useCallback, useState } from 'react'
import { graphql, useFragment } from 'react-relay'

import { AlertMessageType } from 'components/types'
import { makeUpsertTopic, MatchingTopicsType } from 'mutations/upsertTopicMutation'
import { AddTopic_viewer$key } from '__generated__/AddTopic_viewer.graphql'
import { AddTopic_parentTopic$key } from '__generated__/AddTopic_parentTopic.graphql'
import UpsertTopicAlert from './UpsertTopicAlert'

const tooltipText = 'Add a subtopic to this topic. You can click "Edit"\n'
  + 'afterwards if it also belongs under another topic.\n'
  + 'Press "Return" to submit the new topic.'

type Props = {
  disabled?: boolean,
  parentTopic: AddTopic_parentTopic$key,
  viewer: AddTopic_viewer$key,
}

const parentTopicFragment = graphql`
  fragment AddTopic_parentTopic on Topic {
    id
  }
`

const viewerFragment = graphql`
  fragment AddTopic_viewer on User {
    selectedRepoId
  }
`

export default function AddTopic(props: Props) {
  const viewer = useFragment(viewerFragment, props.viewer)
  const parentTopic = useFragment(parentTopicFragment, props.parentTopic)
  const [name, setName] = useState('')

  const selectedRepoId = viewer.selectedRepoId
  if (!selectedRepoId) return null

  const makeAlert = useCallback(
    (alert: AlertMessageType, matchingTopics: MatchingTopicsType) => (
      <UpsertTopicAlert
        alert={alert}
        matchingTopics={matchingTopics}
        name={name}
        parentTopicId={parentTopic.id}
        selectedRepoId={selectedRepoId}
      />
    ),
    [name, selectedRepoId],
  )

  const ignoreEvent = useCallback(
    (event: KeyboardEvent<HTMLInputElement>) => event.key !== 'Enter',
    [],
  )

  const onKeyPress = makeUpsertTopic({
    selectedRepoId, name, setName, parentTopicId: parentTopic.id, makeAlert, onMatchingSynonym: 'ASK',
    ignoreEvent,
  })

  const updateName = useCallback((event: FormEvent<HTMLInputElement>) => {
    setName(event.currentTarget.value)
  }, [setName])

  return (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-ne"
          aria-label={tooltipText}
        >
          <label htmlFor="create-topic-name">Add subtopic</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control test-topic-name input-sm"
          disabled={props.disabled}
          id="create-topic-name"
          onChange={updateName}
          onKeyPress={onKeyPress}
          placeholder="Name or description"
          data-testid="topic-name-input"
          value={name}
        />
      </dd>
    </dl>
  )
}
