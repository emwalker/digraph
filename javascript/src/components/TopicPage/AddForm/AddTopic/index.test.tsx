import React from 'react'
import { shallow } from 'enzyme'
import AddTopic from './index'

jest.mock('react-relay', () => ({ createFragmentContainer: (component: any) => component }))

describe('<AddTopic />', () => {
  // @ts-expect-error
  const wrapper = shallow(<AddTopic />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('includes an input for the name', () => {
    expect(wrapper.find('input').hasClass('test-topic-name')).toBeTruthy()
  })

  it('includes a tooltip', () => {
    expect(wrapper.find('.tooltipped').exists()).toBeTruthy()
  })
})
