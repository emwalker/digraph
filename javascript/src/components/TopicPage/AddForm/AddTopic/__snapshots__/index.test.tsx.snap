// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddTopic /> renders 1`] = `
<dl
  className="form-group"
>
  <dt>
    <span
      aria-label="Add a subtopic to this topic. You can click \\"Edit\\"
afterwards if it also belongs under another topic.
Press \\"Return\\" to submit the new topic."
      className="tooltipped tooltipped-ne"
    >
      <label
        htmlFor="create-topic-name"
      >
        Add subtopic
      </label>
    </span>
  </dt>
  <dd>
    <input
      className="form-control test-topic-name input-sm"
      disabled={true}
      id="create-topic-name"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Name or description"
      value=""
    />
  </dd>
</dl>
`;
